### PR TITLE
[rewarding] add active web3 rewarding settings

### LINKED ISSUE #3731: Add Activation height for web3 rewarding
### Why is this needed
pr3691 follow-up
### How important you think this is for IoTeX
- [] must have
- [] should have
- [] nice to have

### Additional information
Let us know any background or context that would help us better understand the request (for example the particular use-case that prompted this request)

### PR DESCRIPTION
**Pull request by ququzone**
# Description
add active web3 rewarding settings

Fixes #3731 

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [x] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Liuhaai** [comment]
pls refer to https://github.com/iotexproject/iotex-core/commit/39a29ad5894d8a3cede753552c62e95c21f3fdb1#

**Comment by ququzone** [comment]
> pls refer to [39a29ad](https://github.com/iotexproject/iotex-core/commit/39a29ad5894d8a3cede753552c62e95c21f3fdb1)

Thanks, changed.

File: state/factory/workingset.go [hunk 1]
```diff
@@ -34,6 +35,8 @@ var (
 		},
 		[]string{"type"},
 	)
+
+	errUnsupportWeb3Rewarding = errors.New("unsupported web3 rewarding")
```
Comment (envestcc) [inline state/factory/workingset.go:39]: `errors.New` will fill unnecessary stack info in the global defined `errUnsupportWeb3Rewarding `. `fmt.Errorf` maybe better ?

**Review by Liuhaai — APPROVED** [review]

File: state/factory/workingset.go (hunk 1 above)
Comment (dustinxie) [reply state/factory/workingset.go:39]: 1. we used `errors.New` a lot in the codebase, so I think OK to use it to be consistent
2. if we want, we can study this and file a separate PR to replace with `fmt.Errorf`

File: state/factory/workingset.go [hunk 2]
```diff
@@ -576,3 +582,17 @@ func (ws *workingSet) CreateBuilder(
 		SetLogsBloom(calculateLogsBloom(ctx, ws.receipts))
 	return blkBuilder, nil
 }
+
+func isWeb3RewardingAction(selp action.SealedEnvelope) bool {
+	if selp.Encoding() == uint32(iotextypes.Encoding_ETHEREUM_RLP) {
```
Comment (dustinxie) [inline state/factory/workingset.go:587]:
```
if selp.Encoding() != uint32(iotextypes.Encoding_ETHEREUM_RLP) {
    return false
}
xxx
```

File: state/factory/workingset.go [hunk 3]
```diff
@@ -576,3 +582,17 @@ func (ws *workingSet) CreateBuilder(
...
+		act := selp.Action()
+		switch act.(type) {
```
Comment (dustinxie) [inline state/factory/workingset.go:589]: will `switch selp.Action().(type)` compile?

File: state/factory/workingset.go [hunk 4]
```diff
@@ -147,6 +150,9 @@ func (ws *workingSet) runAction(
 	if protocol.MustGetBlockCtx(ctx).GasLimit < protocol.MustGetActionCtx(ctx).IntrinsicGas {
 		return nil, action.ErrGasLimit
 	}
+	if !protocol.MustGetFeatureCtx(ctx).EnableWeb3Rewarding && isWeb3RewardingAction(elp) {
+		return nil, errUnsupportWeb3Rewarding
+	}
```
Comment (dustinxie) [inline state/factory/workingset.go:155]: also add this check into API code, search `validateChainID` you will see how it is used

File: state/factory/workingset.go (hunk 1 above)
Comment (envestcc) [reply state/factory/workingset.go:39]: sure, the consistency is also important. we can study it in a separate issue.

**Review by CoderZhi — APPROVED** [review]

**Comment by Liuhaai** [comment]
Create an issue for the revert of this commit after the hf

**Comment by ququzone** [comment]
> @ququzone Could you assign an issue for yourself to revert this commit after the hf

done #3753

**Review by millken — APPROVED** [review]

**Review by raullenchai — APPROVED** [review]

File: api/coreservice.go [hunk 5]
```diff
@@ -441,6 +446,19 @@ func (core *coreService) validateChainID(chainID uint32) error {
 	return nil
 }
 
+func (core *coreService) validateWeb3Rewarding(selp action.SealedEnvelope) error {
+	if ge := core.bc.Genesis(); ge.IsToBeEnabled(core.bc.TipHeight()) || selp.Encoding() == uint32(iotextypes.Encoding_ETHEREUM_RLP) {
```
Comment (dustinxie) [inline api/coreservice.go:450]: should be `selp.Encoding() == uint32(iotextypes.Encoding_IOTEX_PROTOBUF)` ?

Comment (dustinxie) [reply api/coreservice.go:450]: otherwise a new web3 format rewarding action will return nil here?

File: api/coreservice.go [hunk 6]
```diff
@@ -441,6 +446,19 @@ func (core *coreService) validateChainID(chainID uint32) error {
 	return nil
 }
 
+func (core *coreService) validateWeb3Rewarding(selp action.SealedEnvelope) error {
+	if ge := core.bc.Genesis(); ge.IsToBeEnabled(core.bc.TipHeight()) || selp.Encoding() != uint32(iotextypes.Encoding_ETHEREUM_RLP) {
+		return nil
+	}
+	switch selp.Action().(type) {
+	case *action.ClaimFromRewardingFund,
```
Comment (dustinxie) [inline api/coreservice.go:454]: do i recall correctly, that there's 2 new claim actions enabled by web3 format?
if that's the case, need to check 2 possible types?

**Pull request merged**